### PR TITLE
Add postprocessing for hierarchical classification

### PR DESCRIPTION
**Pull request by sovrasov**
# What does this PR do?

This PR add parsing of hierarchical classification output with further labels resolving.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sovrasov** [comment]
@Wovchena please have a look

File: model_api/python/openvino/model_api/models/classification.py [hunk 1]
```diff
@@ -100,24 +110,71 @@ def parameters(cls):
...
             return self.get_multilabel_predictions(
                 outputs[self.out_layer_names[0]].squeeze()
             )
+        elif self.hierarchical:
+            return self.get_hierarchical_predictions(
+                outputs[self.out_layer_names[0]].squeeze()
+            )
         return self.get_multiclass_predictions(outputs)
 
+    def get_hierarchical_predictions(self, logits: np.ndarray):
```
Comment (Wovchena) [inline model_api/python/openvino/model_api/models/classification.py:139]: Should there be greedy label resolver?

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: model_api/python/openvino/model_api/models/classification.py (hunk 1 above)
Comment (sovrasov) [reply model_api/python/openvino/model_api/models/classification.py:139]: At some point yes, it's just a basic pre-processing, as mentioned in PR's heading

Comment (sovrasov) [reply model_api/python/openvino/model_api/models/classification.py:140]: Added

File: model_api/python/openvino/model_api/models/classification.py [hunk 2]
```diff
@@ -167,3 +226,81 @@ def addOrFindSoftmaxAndTopkOutputs(inference_adapter, topk):
...
+def softmax_numpy(x: np.ndarray, eps: float = 1e-9):
+    x = np.exp(x - np.max(x))
+    return x / (np.sum(x) + eps)
+
+
+class GreedyLabelsResolver:
+    def __init__(self, hierarchical_config) -> None:
+        self.all_labels = list(
+            hierarchical_config["cls_heads_info"]["label_to_idx"].keys()
+        )
```
Comment (Wovchena) [inline model_api/python/openvino/model_api/models/classification.py:240]: That relies on an order of how `keys()` is iterated. I suggest this since you use it as map anyway
```suggestion
        self.idx_to_label = {
            idx: label
            for label, idx, in hierarchical_config["cls_heads_info"][
                "label_to_idx"
            ].items()
        }
        if len(self.idx_to_labeel) != len(
            hierarchical_config["cls_heads_info"]["label_to_idx"]
        ):
            raise RuntimeError(
                'All values in hierarchical_config["cls_heads_info"]["label_to_idx"] must be unique'
            )
```

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: model_api/python/openvino/model_api/models/classification.py (hunk 2 above)
Comment (sovrasov) [reply model_api/python/openvino/model_api/models/classification.py]: I actually need `label_to_idx` itself, just realized it

File: model_api/python/openvino/model_api/models/classification.py [hunk 3]
```diff
@@ -100,24 +111,72 @@ def parameters(cls):
...
+            for i in range(head_logits.shape[0]):
+                if head_logits[i] > self.confidence_threshold:
+                    label_str = cls_heads_info["all_groups"][
+                        cls_heads_info["num_multiclass_heads"] + i
+                    ][0]
+                    predicted_labels.append(label_str)
+                    predicted_indices.append(cls_heads_info["label_to_idx"][label_str])
+                    predicted_scores.append(head_logits[i])
+
+        predictions = list(zip(predicted_indices, predicted_labels, predicted_scores))
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/models/classification.py]: Updated

File: model_api/python/openvino/model_api/models/classification.py
[the remark this answers is not in the document]
Comment (Wovchena) [reply model_api/python/openvino/model_api/models/classification.py:1]: Uploaded. Please, add a line to download the model similar to https://github.com/openvinotoolkit/model_api/blob/67f4fcd555017672dfdd868b6bdc47be04ad0cc2/tests/python/accuracy/prepare_data.py#L59 and add the model with its predictions to https://github.com/openvinotoolkit/model_api/blob/master/tests/python/accuracy/public_scope.json

[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/python/openvino/model_api/models/classification.py:1]: Done

**Review by Wovchena — APPROVED** [review]

**Comment by Wovchena** [comment]
@sovrasov, C++ test failed. Need to revert https://github.com/openvinotoolkit/model_api/blob/master/tests/python/accuracy/public_scope.json for now. Or we can keep the PR hanging until you add C++ version.

**Comment by sovrasov** [comment]
@Wovchena please have a look at C++ part as well

File: model_api/cpp/models/src/classification_model.cpp [hunk 4]
```diff
@@ -184,14 +255,58 @@ std::unique_ptr<ResultBase> ClassificationModel::get_multilabel_predictions(Infe
...
 }
 
+std::unique_ptr<ResultBase> ClassificationModel::get_hierarchical_predictions(InferenceResult& infResult) {
+    const ov::Tensor& logitsTensor = infResult.outputsData.find(outputNames[0])->second;
+    const float* logitsPtr = logitsTensor.data<float>();
+
+    std::vector<std::string> predicted_labels;
+    std::vector<float> predicted_scores;
+    std::vector<float> activated_logits(hierarchical_config.num_single_label_classes);
+    std::copy(logitsPtr, logitsPtr + hierarchical_config.num_single_label_classes, activated_logits.data());
```
Comment (Wovchena) [inline model_api/cpp/models/src/classification_model.cpp:273]:
```suggestion
    std::vector<float> activated_logits(logitsPtr, logitsPtr + hierarchical_config.num_single_label_classes);
```
You only use `activated_logits`'s pointer to data later. Why not use `logitsPtr` directly? Just remove `const` form `const float* logitsPtr = logitsTensor.data<float>();`

File: model_api/cpp/models/include/models/classification_model.h [hunk 5]
```diff
@@ -31,6 +32,34 @@ struct ClassificationResult;
...
 struct ImageInputData;
 
+struct HierarchicalConfig  {
+    std::map<std::string, int> label_to_idx;
+    std::vector<std::pair<std::string, std::string>> label_tree_edges;
+    std::vector<std::vector<std::string>> all_groups;
+    std::map<int, std::pair<int,int>> head_idx_to_logits_range;
+    int num_multiclass_heads;
+    int num_multilabel_heads;
+    int num_single_label_classes;
```
Comment (Wovchena) [inline model_api/cpp/models/include/models/classification_model.h:42]: Can you move from ints to size_t? I hope this will make json validate that the values are positive.

File: model_api/cpp/models/src/classification_model.cpp [hunk 6]
```diff
@@ -184,14 +255,58 @@ std::unique_ptr<ResultBase> ClassificationModel::get_multilabel_predictions(Infe
...
     }
 
     return retVal;
 }
 
+std::unique_ptr<ResultBase> ClassificationModel::get_hierarchical_predictions(InferenceResult& infResult) {
+    const ov::Tensor& logitsTensor = infResult.outputsData.find(outputNames[0])->second;
+    const float* logitsPtr = logitsTensor.data<float>();
+
+    std::vector<std::string> predicted_labels;
```
Comment (Wovchena) [inline model_api/cpp/models/src/classification_model.cpp:270]:
```suggestion
    std::vector<std::reference_wrapper<std::string>> predicted_labels;
```

File: model_api/cpp/models/include/models/classification_model.h [hunk 7]
```diff
@@ -31,6 +32,34 @@ struct ClassificationResult;
 struct ResultBase;
 struct ImageInputData;
 
+struct HierarchicalConfig  {
```
Comment (Wovchena) [inline model_api/cpp/models/include/models/classification_model.h:35]:
```suggestion
struct HierarchicalConfig {
```

File: model_api/cpp/models/src/classification_model.cpp [hunk 8]
```diff
@@ -87,16 +134,31 @@ ClassificationModel::ClassificationModel(std::shared_ptr<ov::Model>& model, cons
...
+
+    auto config_iter = configuration.find("hierarchical_config");
+    if (config_iter == configuration.end()) {
+        if (model->has_rt_info("model_info", "hierarchical_config")) {
+            hierarchical_json_config = model->get_rt_info<std::string>("model_info", "hierarchical_config");
+        }
+    } else {
+        hierarchical_json_config = thresh_iter->second.as<std::string>();
     }
+    if (hierarchical_json_config.size())  {
```
Comment (Wovchena) [inline model_api/cpp/models/src/classification_model.cpp:158]: You should try to read it only `if hierarchical`. But `if (hierarchical && hierarchical_json_config.empty())` this should be an error.

File: model_api/python/openvino/model_api/models/classification.py [hunk 9]
```diff
@@ -32,6 +34,15 @@ def __init__(self, inference_adapter, configuration=None, preload=False):
         if self.path_to_labels:
             self.labels = self._load_labels(self.path_to_labels)
         self.out_layer_names = [self._get_output()]
+
+        if self.hierarchical:
+            self.embedded_processing = True
+            self.hierarchical_info = json.loads(self.hierarchical_config)
```
Comment (Wovchena) [inline model_api/python/openvino/model_api/models/classification.py:40]: Can you throw `if self.hierarchical and not self.hierarchical_config` after my C++ comment

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: model_api/cpp/CMakeLists.txt [hunk 10]
```diff
@@ -28,6 +28,14 @@ set(model_api_VERSION 0.0.0)
 find_package(OpenCV REQUIRED COMPONENTS core imgproc)
 find_package(OpenVINO REQUIRED COMPONENTS Runtime)
 
+if(WIN32)
+    include(FetchContent)
+    FetchContent_Declare(json URL https://github.com/nlohmann/json/releases/download/v3.11.2/json.tar.xz)
+    FetchContent_MakeAvailable(json)
+else()
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/cpp/CMakeLists.txt:35]: nlohmann-json3-dev package is required by openvino, so I decided to use find_pakcage for json. Do we support something different than debian/ubuntu?  I forgot to add nlohmann-json3-dev to install_dependencies.sh

[the remark this answers is not in the document]
Comment (Wovchena) [reply model_api/cpp/CMakeLists.txt:35]: > Do we support something different than debian/ubuntu?

No. (if to exclude Windows)

> nlohmann-json3-dev package is required by openvino

Strange. They also have a submodule https://github.com/openvinotoolkit/openvino/tree/master/thirdparty/json

[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/cpp/CMakeLists.txt:35]: I also was surprised, maybe the relation is not direct: if you try `sudo apt remove nlohmann-json3-dev`, you'll get 
```
The following packages will be REMOVED:
  nlohmann-json3-dev openvino openvino-2022.3.0 openvino-samples-2022.3.0
```

[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/cpp/CMakeLists.txt:35]: Are we good with updating install_dependencies.sh or I should delete find_package call?

File: model_api/cpp/CMakeLists.txt [hunk 11]
```diff
@@ -47,6 +55,7 @@ source_group("adapters/include" FILES ${ADAPTERS_HEADERS})
 add_library(model_api STATIC ${MODELS_SOURCES} ${UTILS_SOURCES} ${ADAPTERS_SOURCES})
 target_include_directories(model_api PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/models/include>" "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/utils/include>" "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/adapters/include>" "$<INSTALL_INTERFACE:include>")
 target_link_libraries(model_api PUBLIC openvino::runtime opencv_core opencv_imgproc)
+target_link_libraries(model_api PRIVATE $<BUILD_LOCAL_INTERFACE:nlohmann_json::nlohmann_json>)
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/cpp/CMakeLists.txt:58]: Actually, I've been trying to workaround this, but with no luck. Just PRIVATE is ok (like in tests), unless we call export()

[the remark this answers is not in the document]
Comment (Wovchena) [reply model_api/cpp/CMakeLists.txt:58]: Well, update cmake version then

File: model_api/cpp/models/src/classification_model.cpp (hunk 8 above)
Comment (sovrasov) [reply model_api/cpp/models/src/classification_model.cpp]: Done

File: model_api/cpp/models/include/models/classification_model.h (hunk 7 above)
Comment (sovrasov) [reply model_api/cpp/models/include/models/classification_model.h]: Fixed

File: model_api/cpp/models/include/models/classification_model.h (hunk 5 above)
Comment (sovrasov) [reply model_api/cpp/models/include/models/classification_model.h]: Done

File: model_api/cpp/models/include/models/classification_model.h [hunk 12]
```diff
@@ -31,6 +32,34 @@ struct ClassificationResult;
...
+    int num_multilabel_heads;
+    int num_single_label_classes;
+
+    HierarchicalConfig();
+    HierarchicalConfig(const std::string&);
+};
+
+class GreedyLabelsResolver {
+    public:
+        GreedyLabelsResolver();
```
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/cpp/models/include/models/classification_model.h]: Fixed

File: model_api/cpp/CMakeLists.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (sovrasov) [reply model_api/cpp/CMakeLists.txt:54]: Updated

**Review by Wovchena — APPROVED** [review]

**Comment by sovrasov** [comment]
There are 2 CI failures:
resnet-18-pytorch returns unexpected result: test_data["reference"][i]='(222, kuvasz, 0.140)', output_str='(254, pug, 0.153)'
and mobilenet-v3-large-1.0-224-tf fails to be converted into IR: [ ERROR ]  No node with name StatefulPartitionedCall/MobilenetV3large/Predictions/Softmax

Is there any chance that some of the assets were silently updated (I see that the test packages versions in setup.py are not fixes, 
 OMZ models could also be updated)? I can't reproduce python failures in my local environment (but it's almost two weeks odl).

**Comment by Wovchena** [comment]
There were no changes for `omz_tools` (which is actually part of openvino-dev) but openvino-dev 2023.0.0 just got published as well as openvino itself.

I guess resnet-18-pytorch's error is specific to openvino version. I will update the reference after C++ package is published.

Is probably OMZ's fault in combination with new openvino.

Do you have urgency to merge this PR? I can merge it now, but I would prefer to fix these issues first.

**Comment by sovrasov** [comment]
> There were no changes for `omz_tools` (which is actually part of openvino-dev) but openvino-dev 2023.0.0 just got published as well as openvino itself.
> 
> I guess resnet-18-pytorch's error is specific to openvino version. I will update the reference after C++ package is published.
> 
> Is probably OMZ's fault in combination with new openvino.
> 
> Do you have urgency to merge this PR? I can merge it now, but I would prefer to fix these issues first.

No problem, let's wait for the final OV update.

File: model_api/cpp/models/src/classification_model.cpp (hunk 4 above)
Comment (Wovchena) [reply model_api/cpp/models/src/classification_model.cpp]: I didn't pay attention to this at the moment when I approved. What about
> You only use `activated_logits`'s pointer to data later. Why not use `logitsPtr` directly? Just remove `const` form `const float* logitsPtr = logitsTensor.data<float>();`

Comment (sovrasov) [reply model_api/cpp/models/src/classification_model.cpp]: I thought you mean avoiding direct copy call and use constructor of activated_logits for that
